data-prepper-core/src/main/java/org/opensearch/dataprepper/parser/CircuitBreakingBuffer.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.parser;
+
+import org.opensearch.dataprepper.breaker.CircuitBreaker;
+import org.opensearch.dataprepper.model.CheckpointState;
+import org.opensearch.dataprepper.model.buffer.Buffer;
+import org.opensearch.dataprepper.model.record.Record;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.TimeoutException;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Decorator for {@link Buffer} which checks a {@link CircuitBreaker}
+ * before writing records.
+ *
+ * @param <T> The type of record.
+ * @since 2.1
+ */
+class CircuitBreakingBuffer<T extends Record<?>> implements Buffer<T> {
+    private final Buffer<T> buffer;
+    private final CircuitBreaker circuitBreaker;
+
+    /**
+     * Constructor
+     *
+     * @param buffer The inner buffer which is being decorated
+     * @param circuitBreaker The circuit breaker to check
+     */
+    public CircuitBreakingBuffer(final Buffer<T> buffer, final CircuitBreaker circuitBreaker) {
+        this.buffer = requireNonNull(buffer);
+        this.circuitBreaker = requireNonNull(circuitBreaker);
+    }
+
+    @Override
+    public void write(final T record, final int timeoutInMillis) throws TimeoutException {
+        checkBreaker();
+
+        buffer.write(record, timeoutInMillis);
+    }
+
+    @Override
+    public void writeAll(final Collection<T> records, final int timeoutInMillis) throws Exception {
+        checkBreaker();
+
+        buffer.writeAll(records, timeoutInMillis);
+    }
+
+    private void checkBreaker() throws TimeoutException {
+        if(circuitBreaker.isOpen())
+            throw new TimeoutException("Circuit breaker is open. Unable to write to buffer.");
```

Comment:
Should this be a custom exception, maybe CircuitBreakingException?

Reply:
I used the `TimeoutException` because that is what `Buffer` currently throws for such situations. Callers of `Buffer` already have some expectation here on what it throws. Maybe it won't affect existing callers to add a new exception. I could check usages.

Reply:
My thought was that callers may want to distinguish between the circuit breaker being tripped vs other timeout scenarios to decide on what action should be taken. If adding the new exception would require a bunch of refactors to the callers then it's probably not worth it for now

Reply:
One other thought on this is that I tend to think the sources (which write to buffers) probably should not be too aware of the different reasons the buffer can fill.

If want some form of global action, we could add that here. 

I propose leaving this as it is currently written. In the future if we have a concrete need, we can discuss this idea further and make the change.